rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,563 @@
+
+Any Cloud tenants which are lagging behind the current release and are required to jump several versions to land on the
+current release should revert to using recreate instead of a rolling deploy. We can also add a flag that indicates
+migrations should be performed during initialization instead of in the background like during the phased approach. While
+this will negate the changes described in this RFD to eliminate downtime, this kind of jump in releases should be the
+exception and not the norm.
+
+### Major/Breaking changes
+
+For major breaking changes (i.e., splitting a resource into multiple resources), resources must be migrated to a new
+key. For example, to migrate the data stored in `some/path/to/resource/<ID>` we must leave the original value at
+`some/path/to/resource/<ID>` as is, and write the migrated value to `some/new/path/to/resource/<ID>`. Keys should be
+versioned to determine which version of a resource exists at any given key range. A key prefix of
+`/type/version/subkind/name` should be used where possible to have uniformity. For example, if nodes were migrated from
+`/nodes/default/<UUID>` it should be to `/nodes/v2/default/<UUID>`.
+
+Any breaking changes to a resource must also bump the API version of the service which interacts with that resource to
+ensure existing code does not break and to make it clear when new clients are choosing to use the new behavior.
+
+To perform the migration which allows for backward compatibility with the previous version, each step of the migration
+will be done in a phase. Each phase can be gated by a feature flag which is updated to the next phase in the process on
+subsequent Cloud releases. When all the phases have been completed the code can be cleaned up to remove the transitional
+state and feature flags required to get to the target state. The compatibility guarantees of this approach do come with
+a cost though: it will now take longer and require more process to perform the migration. We can however mitigate the
+time required to progress through each phase by increasing the Cloud release cadence.
+
+| Phase | Write                  | Read                                        | Details                                                                       |
+| ----- | ---------------------- | ------------------------------------------- | ----------------------------------------------------------------------------- |
+| 1     | Old key AND new key(s) | Old Key                                     | Introduce new resource(s) and mirror data to new key range on write           |
+| 2     | Old key AND new key(s) | New Key, fallback to old key if not present | Start reading from the new key and using the old key as a fall back           |
+| 3     | Old key AND new key(s) | New Key, fallback to old key if not present | Run data migration that copies all keys from old to new                       |
```

Comment:
I don't understand this step. Wouldn't the "migration that copies all keys from old to new" have the potential to overwrite valid data in the new keys that was written during phase 1 or 2?